java/arcs/core/analysis/RecipeGraph.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.analysis
+
+import arcs.core.data.HandleConnectionSpec
+import arcs.core.data.HandleMode
+import arcs.core.data.Recipe
+
+/**
+ * A graph capturing the connections between particles and handles in [recipe].
+ *
+ * Nodes in the graph are either a [Recipe.Handle] or [Recipe.Particle]. For every write connection
+ * from a particle `p` to a handle `h` using a connection spec `s`, there is a labeled edge
+ * `p -s-> h` in the graph. Similarly, for every read connection from a particle `p` to a handle `h`
+ * using a connection spec `s`, there is a labeled edge `h -s-> p` in the graph.
+*/
+class RecipeGraph(recipe: Recipe) {
+    var nodes = mutableListOf<Node>()
```

Comment:
Maybe something like:
```
class RecipeGraph(recipe: Recipe) {
    val nodes: List<Node> = calculateRecipeNodes(recipe)
    (...)
}
```

Reply:
(to avoid public mutable variable)

Reply:
Can the field itself be made private, and instead - introduce an API with more useful/understandable semantics for the intended use-case?

Reply:
I would mostly need to iterate on the nodes.I made this into an immutable list.